File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.py
# Time complexity: O(n). Only a single pass is needed.

# Space complexity: O(1). Only two variables are used.
    
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0
        
        for i in range(len(prices)):
            if prices[i] < min_price:
                min_price = prices[i]
            elif prices[i] - min_price > max_profit:
                max_profit = prices[i] - min_price
        
        return max_profit
        

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.js
// Approach: One Pass

// Time complexity: O(n). Only a single pass is needed.
// Space complexity: O(1). Only two variables are used.

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let min_price = Infinity;
    let max_profit = 0;
    
    for (let i = 0; i <prices.length; i++) {
        if (prices[i] < min_price) {
            min_price = prices[i];
        } else if (prices[i] - min_price > max_profit) {
            max_profit = prices[i] - min_price;
        }
    }    
    return max_profit;
};